File: examples/python/basic_usage.py
"""
Basic usage example of the Miktos API with Python.
"""
import os
import requests

# API configuration
API_BASE_URL = "https://api.miktos.ai/v1"
API_KEY = os.environ.get("MIKTOS_API_KEY", "your-api-key-here")

class MiktosClient:
    """Simple Python client for the Miktos API."""
    
    def __init__(self, api_key, base_url=API_BASE_URL):
        self.api_key = api_key
        self.base_url = base_url
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
    
    def authenticate(self, email, password):
        """Get an authentication token."""
        response = requests.post(
            f"{self.base_url}/auth/token",
            json={"email": email, "password": password}
        )
        response.raise_for_status()
        data = response.json()
        self.api_key = data["access_token"]
        self.headers["Authorization"] = f"Bearer {self.api_key}"
        return data
    
    def list_projects(self, skip=0, limit=100):
        """List all user projects."""
        response = requests.get(
            f"{self.base_url}/projects",
            headers=self.headers,
            params={"skip": skip, "limit": limit}
        )
        response.raise_for_status()
        return response.json()
    
    def create_project(self, name, description=None, context_notes=None, repository_url=None):
        """Create a new project."""
        data = {
            "name": name,
            "description": description,
            "context_notes": context_notes,
            "repository_url": repository_url
        }
        # Remove None values
        data = {k: v for k, v in data.items() if v is not None}
        
        response = requests.post(
            f"{self.base_url}/projects",
            headers=self.headers,
            json=data
        )
        response.raise_for_status()
        return response.json()
    
    def generate_text(self, project_id, model, messages, temperature=0.7, max_tokens=1024):
        """Generate text using the specified model."""
        data = {
            "project_id": project_id,
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens
        }
        
        response = requests.post(
            f"{self.base_url}/generate",
            headers=self.headers,
            json=data
        )
        response.raise_for_status()
        return response.json()
    
    def generate_text_stream(self, project_id, model, messages, temperature=0.7, max_tokens=1024):
        """Generate text using the specified model with streaming."""
        data = {
            "project_id": project_id,
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": True
        }
        
        with requests.post(
            f"{self.base_url}/generate",
            headers=self.headers,
            json=data,
            stream=True
        ) as response:
            response.raise_for_status()
            for line in response.iter_lines():
                if line:
                    # Process the streamed response 
                    # In a real implementation, you would parse the SSE format
                    yield line.decode('utf-8')


# Example usage
def main():
    client = MiktosClient(API_KEY)
    
    # Creating a new project
    project = client.create_project(
        name="Python API Example",
        description="Testing the Miktos API with Python"
    )
    print(f"Created project: {project['name']} with ID: {project['id']}")
    
    # Generating text
    result = client.generate_text(
        project_id=project['id'],
        model="openai/gpt-4o",
        messages=[
            {"role": "user", "content": "Explain quantum computing in simple terms."}
        ],
        temperature=0.7
    )
    
    print("\nGenerated response:")
    print(result['content'])
    
    # List all projects
    projects = client.list_projects()
    print(f"\nTotal projects: {len(projects)}")
    
    # Example of streaming (just printing raw output here)
    print("\nStreaming example (raw output):")
    for chunk in client.generate_text_stream(
        project_id=project['id'],
        model="anthropic/claude-3-opus",
        messages=[
            {"role": "user", "content": "Write a short poem about AI."}
        ]
    ):
        print(chunk)


if __name__ == "__main__":
    main()

File: examples/javascript/basic_usage.js
/**
 * Basic usage example of the Miktos API with Node.js
 */

const fetch = require('node-fetch');

// API configuration
const API_BASE_URL = 'https://api.miktos.ai/v1';
const API_KEY = process.env.MIKTOS_API_KEY || 'your-api-key-here';

class MiktosClient {
  constructor(apiKey, baseUrl = API_BASE_URL) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
    this.headers = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json'
    };
  }

  /**
   * Get an authentication token
   * @param {string} email - User email
   * @param {string} password - User password
   * @returns {Promise<Object>} Auth token response
   */
  async authenticate(email, password) {
    const response = await fetch(`${this.baseUrl}/auth/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password })
    });
    
    if (!response.ok) {
      throw new Error(`Authentication failed with status: ${response.status}`);
    }
    
    const data = await response.json();
    this.apiKey = data.access_token;
    this.headers['Authorization'] = `Bearer ${this.apiKey}`;
    return data;
  }

  /**
   * List all user projects
   * @param {number} skip - Number of projects to skip
   * @param {number} limit - Maximum number of projects to return
   * @returns {Promise<Array>} List of projects
   */
  async listProjects(skip = 0, limit = 100) {
    const response = await fetch(
      `${this.baseUrl}/projects?skip=${skip}&limit=${limit}`,
      { headers: this.headers }
    );
    
    if (!response.ok) {
      throw new Error(`Failed to list projects with status: ${response.status}`);
    }
    
    return response.json();
  }

  /**
   * Create a new project
   * @param {string} name - Project name
   * @param {string} description - Project description
   * @param {string} contextNotes - Additional context notes
   * @param {string} repositoryUrl - Git repository URL
   * @returns {Promise<Object>} Created project
   */
  async createProject(name, description = null, contextNotes = null, repositoryUrl = null) {
    const data = {
      name,
      description,
      context_notes: contextNotes,
      repository_url: repositoryUrl
    };
    
    // Remove null values
    Object.keys(data).forEach(key => {
      if (data[key] === null) {
        delete data[key];
      }
    });
    
    const response = await fetch(`${this.baseUrl}/projects`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(data)
    });
    
    if (!response.ok) {
      throw new Error(`Failed to create project with status: ${response.status}`);
    }
    
    return response.json();
  }

  /**
   * Generate text using the specified model
   * @param {string} projectId - Project ID
   * @param {string} model - Model identifier (e.g., "openai/gpt-4o")
   * @param {Array} messages - Array of message objects with role and content
   * @param {number} temperature - Sampling temperature
   * @param {number} maxTokens - Maximum tokens to generate
   * @returns {Promise<Object>} Generation result
   */
  async generateText(projectId, model, messages, temperature = 0.7, maxTokens = 1024) {
    const data = {
      project_id: projectId,
      model,
      messages,
      temperature,
      max_tokens: maxTokens
    };
    
    const response = await fetch(`${this.baseUrl}/generate`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(data)
    });
    
    if (!response.ok) {
      throw new Error(`Text generation failed with status: ${response.status}`);
    }
    
    return response.json();
  }

  /**
   * Generate text using the specified model with streaming
   * @param {string} projectId - Project ID
   * @param {string} model - Model identifier (e.g., "openai/gpt-4o")
   * @param {Array} messages - Array of message objects with role and content
   * @param {number} temperature - Sampling temperature
   * @param {number} maxTokens - Maximum tokens to generate
   * @param {function} onChunk - Callback function for each chunk
   * @returns {Promise<void>}
   */
  async generateTextStream(projectId, model, messages, onChunk, temperature = 0.7, maxTokens = 1024) {
    const data = {
      project_id: projectId,
      model,
      messages,
      temperature,
      max_tokens: maxTokens,
      stream: true
    };
    
    const response = await fetch(`${this.baseUrl}/generate`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(data)
    });
    
    if (!response.ok) {
      throw new Error(`Text generation failed with status: ${response.status}`);
    }
    
    // Process the response as a stream
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      
      const chunk = decoder.decode(value);
      onChunk(chunk);
    }
  }
}

// Example usage
async function main() {
  const client = new MiktosClient(API_KEY);
  
  try {
    // Creating a new project
    const project = await client.createProject(
      'JavaScript API Example',
      'Testing the Miktos API with Node.js'
    );
    console.log(`Created project: ${project.name} with ID: ${project.id}`);
    
    // Generating text
    const result = await client.generateText(
      project.id,
      'openai/gpt-4o',
      [
        { role: 'user', content: 'Explain quantum computing in simple terms.' }
      ],
      0.7
    );
    
    console.log('\nGenerated response:');
    console.log(result.content);
    
    // List all projects
    const projects = await client.listProjects();
    console.log(`\nTotal projects: ${projects.length}`);
    
    // Example of streaming
    console.log('\nStreaming example:');
    await client.generateTextStream(
      project.id,
      'anthropic/claude-3-opus',
      [
        { role: 'user', content: 'Write a short poem about AI.' }
      ],
      (chunk) => console.log(chunk),
      0.7
    );
  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run the example if this file is executed directly
if (require.main === module) {
  main();
}

module.exports = MiktosClient;
